File: engine/thumbnail-generator.js
/**
 * Thumbnail Generator (Node.js)
 *
 * - situation-first
 * - seeded RNG (같은 입력 → 같은 결과)
 * - 풍부한 출력: 폰트 느낌, 피해야 할 요소, Unsplash/Pexels 검색 키워드
 */

"use strict";

const path = require("path");
const fs = require("fs");
const { pickCompatibleMood, filterVisualsForSituation, getSituationVisuals, SITUATION_VISUAL_BLOCK } = require("./coherence");
const { resolveGenre, resolveSituation } = require("./util/aliases");
const { makeRng, pickOne } = require("./util/seeded-random");

let _dict = null;
function loadDict() {
  if (!_dict) {
    _dict = JSON.parse(
      fs.readFileSync(path.join(__dirname, "data", "keyword_dictionary.json"), "utf-8")
    );
  }
  return _dict;
}

// ── 폰트 느낌 (situation × mood) ──
const FONT_FEEL = {
  workout:    "굵은 산세리프 + 대문자, 강한 무게감",
  party:      "굵은 디스플레이 + 네온 효과",
  study:      "얇은 산세리프 + 여백, 미니멀",
  sleep:      "얇은 세리프 + 큰 자간, 부드러움",
  cafe:       "필기체 또는 따뜻한 세리프, 손글씨 느낌",
  night_drive:"콘덴스드 산세리프 + 시네마틱",
  late_night: "콘덴스드 세리프 + 글로우, 미스터리",
  rain:       "얇은 세리프 + 손글씨, 차분함",
  morning:    "라운드 산세리프, 밝고 부드러움",
  walk:       "라운드 산세리프 + 캐주얼",
  commute:    "산세리프 + 또렷한 자간",
  reading:    "북 세리프, 클래식한 글씨",
  cooking:    "라운드 산세리프, 따뜻한 톤",
  breakup:    "얇은 세리프 + 여백, 멜랑꼴리",
  sunset:     "세리프 + 따뜻한 톤",
  travel:     "굵은 산세리프 + 여행 폴라로이드 느낌",
};

// ── 피해야 할 요소 (situation 기반) ──
function _avoidList(sit) {
  const block = SITUATION_VISUAL_BLOCK[sit];
  if (!block || !block.size) {
    return ["과도한 텍스트", "낚시성 자극 이미지", "왜곡된 인물"];
  }
  const arr = [...block].slice(0, 6);
  return ["과도한 텍스트 오버레이", ...arr];
}

// ── Unsplash/Pexels 검색 키워드 ──
function _photoKeywords(genre, mood, sit) {
  const d = loadDict();
  const sitData = (d.situations || {})[resolveSituation(sit)] || {};
  const moodData = (d.moods || {})[mood] || {};
  const genreData = (d.genres || {})[resolveGenre(genre)] || {};
  const sitVis = sitData.visual_keywords || [];
  const moodVis = moodData.visual_concepts || [];
  const genreVis = genreData.visual_keywords || [];
  const sitVisuals = getSituationVisuals(resolveSituation(sit));

  const all = [
    ...(sitVis.slice(0, 3)),
    ...(moodVis.slice(0, 2)),
    ...(genreVis.slice(0, 2)),
    ...(sitVisuals.slice(0, 2)),
  ];
  const seen = new Set();
  const out = [];
  for (const v of all) {
    const k = (v || "").toLowerCase();
    if (!k || seen.has(k)) continue;
    seen.add(k);
    out.push(v);
    if (out.length >= 6) break;
  }
  return out;
}

// ── 컬러톤 ──
function _getColorTones(genre, mood, situation) {
  const d = loadDict();
  const gKey = resolveGenre(genre);
  const sKey = resolveSituation(situation);
  let tones = [];
  for (const [, scene] of Object.entries((d.visuals || {}).visual_scenes || {})) {
    if ((scene.best_for_situations || []).includes(sKey) && (scene.best_for_moods || []).includes(mood)) {
      tones.push(...(scene.color_palette || []));
      break;
    }
  }
  if (!tones.length) {
    tones.push(...((d.genres || {})[gKey] || {}).color_tones || []);
    tones.push(...((d.moods || {})[mood] || {}).color_tones || []);
    tones.push(...((d.situations || {})[sKey] || {}).color_tones || []);
  }
  if (!tones.length) tones = ["dark blue", "warm orange", "soft gray"];
  return [...new Set(tones)].slice(0, 4);
}

// ── 비주얼 컨셉 ──
function _getVisuals(mood, situation) {
  const sitVis = getSituationVisuals(resolveSituation(situation));
  const d = loadDict();
  const moodVis = ((d.moods || {})[mood] || {}).visual_concepts || [];
  const compat = filterVisualsForSituation(resolveSituation(situation), moodVis);
  const combined = [...sitVis];
  for (const v of compat) if (!combined.includes(v)) combined.push(v);
  return combined.slice(0, 6);
}

// ── 인물 여부 ──
function _shouldHavePerson(mood, sit) {
  const s = resolveSituation(sit);
  if (["workout", "party", "commute", "walk"].includes(s)) return true;
  if (["study", "sleep", "reading", "rain"].includes(s)) return false;
  return ["sexy", "romantic", "emotional", "happy", "energetic", "intense"].includes(mood);
}

// ── 레이아웃 (deterministic) ──
function _getLayout(sit, rng) {
  const d = loadDict();
  const layouts = d.thumbnail_layouts || ["center portrait + bold serif title"];
  const pref = {
    workout: "full bleed background + overlay text bottom",
    party: "collage grid 2x2 with overlay title",
    study: "minimal center text on blurred background",
    sleep: "minimal center text on blurred background",
    cafe: "left portrait / right text",
  };
  const sKey = resolveSituation(sit);
  if (pref[sKey] && layouts.includes(pref[sKey])) return pref[sKey];
  return pickOne(rng, layouts);
}

// ── 텍스트 오버레이 (deterministic) ──
const OV_KO = ["{m} {g}", "{s} 감성", "{m} Playlist", "{g} Mix", "{s} Mood", "{m} Vibes"];
const OV_EN = ["{m} {g}", "{s} Vibes", "{m} Playlist", "{g} Mix", "{s} Mood", "Feel the {m}"];

function _overlay(genre, mood, sit, lang, rng) {
  const d = loadDict();
  const gk = (d.genres || {})[resolveGenre(genre)] || {};
  const mk = (d.moods || {})[mood] || {};
  const sk = (d.situations || {})[resolveSituation(sit)] || {};
  const [g, m, s] = lang === "ko"
    ? [(gk.ko_keywords || [genre])[0], (mk.ko_keywords || [mood])[0], (sk.ko_keywords || [sit])[0]]
    : [(gk.en_keywords || [genre])[0], (mk.en_keywords || [mood])[0], (sk.en_keywords || [sit])[0]];
  const tmpls = lang === "ko" ? OV_KO : OV_EN;
  const t = pickOne(rng, tmpls);
  return t.replace(/\{g\}/g, g).replace(/\{m\}/g, m).replace(/\{s\}/g, s);
}

/**
 * Build a thumbnail concept deterministically.
 *
 * @param {object} analysis
 * @param {string} language
 * @param {number} seed uint32 — same seed produces the same result
 * @returns {object} ThumbnailSuggestion
 */
function generateThumbnail(analysis, language, seed) {
  const rng = makeRng((seed >>> 0) || 1);
  const genre = analysis.primaryGenre;
  const sit = analysis.primarySituation;
  const mood = pickCompatibleMood(resolveSituation(sit), analysis.detectedMoods);
  const d = loadDict();
  const gKey = resolveGenre(genre);
  const sKey = resolveSituation(sit);
  const sitEn = ((d.situations || {})[sKey] || {}).en_keywords || [sKey];
  const moodEn = ((d.moods || {})[mood] || {}).en_keywords || [mood];
  const genreEn = ((d.genres || {})[gKey] || {}).en_keywords || [gKey];

  const concepts = _getVisuals(mood, sit);
  const tones = _getColorTones(genre, mood, sit);
  const photoKw = _photoKeywords(genre, mood, sit);

  const mainKw = [sitEn[0], moodEn[0], genreEn[0], concepts[0] || ""].filter(Boolean).slice(0, 5);
  const subKw = [...concepts.slice(1, 4), ...tones.slice(0, 2)].slice(0, 5);

  return {
    mainKeywords: mainKw,
    subKeywords: subKw,
    colorTone: tones,
    backgroundConcept: concepts[0] || "abstract gradient",
    composition: concepts[1] || (concepts[0] ? `centered ${concepts[0]}` : "rule of thirds composition"),
    hasPerson: _shouldHavePerson(mood, sit),
    layout: _getLayout(sit, rng),
    textOverlay: _overlay(genre, mood, sit, language, rng),
    fontFeel: FONT_FEEL[sKey] || "산세리프 + 미디엄 웨이트, 균형감",
    avoidList: _avoidList(sKey),
    photoSearchKeywords: photoKw,
  };
}

module.exports = { generateThumbnail };

File: backend/app/services/thumbnail_generator.py
"""
썸네일 키워드 / 시안 생성 엔진

출력:
- 메인 검색 키워드
- 보조 키워드
- 컬러톤
- 배경 콘셉트
- 인물 여부
- 레이아웃 제안
- 텍스트 오버레이 문구
"""

import json
import random
from pathlib import Path

from backend.app.models.schemas import AnalysisResult, ThumbnailSuggestion

_DATA_DIR = Path(__file__).resolve().parent.parent / "data"
_dict_cache: dict | None = None


def _load_dictionary() -> dict:
    global _dict_cache
    if _dict_cache is None:
        with open(_DATA_DIR / "keyword_dictionary.json", encoding="utf-8") as f:
            _dict_cache = json.load(f)
    return _dict_cache


# ── 컬러톤 ──


def _get_color_tones(genre: str, mood: str) -> list[str]:
    d = _load_dictionary()
    tones: list[str] = []

    genre_tones = d.get("genres", {}).get(genre, {}).get("color_tones", [])
    mood_tones = d.get("moods", {}).get(mood, {}).get("color_tones", [])
    tones.extend(genre_tones)
    tones.extend(mood_tones)

    if not tones:
        tones = ["dark blue", "warm orange", "soft gray"]

    # 중복 제거
    seen: set[str] = set()
    unique: list[str] = []
    for t in tones:
        if t.lower() not in seen:
            seen.add(t.lower())
            unique.append(t)
    return unique[:4]


# ── 비주얼 콘셉트 ──


def _get_visual_concepts(mood: str, situation: str) -> list[str]:
    d = _load_dictionary()
    concepts: list[str] = []

    mood_vis = d.get("moods", {}).get(mood, {}).get("visual_concepts", [])
    sit_vis = d.get("situations", {}).get(situation, {}).get("visual_concepts", [])
    concepts.extend(mood_vis)
    concepts.extend(sit_vis)

    if not concepts:
        concepts = ["city night", "sunset", "headphones"]
    return list(dict.fromkeys(concepts))[:6]


# ── 레이아웃 ──


def _get_layout(mood: str) -> str:
    d = _load_dictionary()
    layouts = d.get("thumbnail_layouts", [])
    if not layouts:
        layouts = [
            "center portrait + bold serif title",
            "left portrait / right text",
            "full bleed background + overlay text",
            "split screen with gradient",
        ]
    return random.choice(layouts)


# ── 인물 여부 추정 ──


_PERSON_MOODS = {"sexy", "romantic", "emotional", "happy", "energetic", "intense"}
_NO_PERSON_MOODS = {"peaceful", "dreamy", "chill"}


def _should_have_person(mood: str, situation: str) -> bool:
    if mood in _PERSON_MOODS:
        return True
    if mood in _NO_PERSON_MOODS and situation in {"sleep", "study", "reading"}:
        return False
    return random.choice([True, False])


# ── 메인/보조 키워드 ──


def _build_main_keywords(
    genre: str, mood: str, situation: str, language: str
) -> list[str]:
    """영문 검색용 메인 키워드 (이미지 검색 최적화)."""
    d = _load_dictionary()
    genre_en = d.get("genres", {}).get(genre, {}).get("en_keywords", [genre])
    mood_en = d.get("moods", {}).get(mood, {}).get("en_keywords", [mood])
    sit_en = d.get("situations", {}).get(situation, {}).get("en_keywords", [situation])

    main = []
    if sit_en:
        main.append(sit_en[0])
    if mood_en:
        main.append(mood_en[0])
    if genre_en:
        main.append(genre_en[0])

    # 콘셉트 단어 추가
    concepts = _get_visual_concepts(mood, situation)
    if concepts:
        main.append(concepts[0])

    return main[:5]


def _build_sub_keywords(
    genre: str, mood: str, situation: str
) -> list[str]:
    concepts = _get_visual_concepts(mood, situation)
    tones = _get_color_tones(genre, mood)

    sub: list[str] = []
    sub.extend(concepts[1:4])
    sub.extend(tones[:2])
    return sub[:5]


# ── 텍스트 오버레이 ──

_OVERLAY_KO_TEMPLATES = [
    "{mood} {genre}",
    "{situation} 감성",
    "{mood} Playlist",
    "{genre} Mix",
    "{situation} Mood",
    "{mood} Vibes",
]

_OVERLAY_EN_TEMPLATES = [
    "{mood} {genre}",
    "{situation} Vibes",
    "{mood} Playlist",
    "{genre} Mix",
    "{situation} Mood",
    "Feel the {mood}",
]


def _generate_overlay_text(
    genre: str, mood: str, situation: str, language: str
) -> str:
    d = _load_dictionary()
    genre_kw = d.get("genres", {}).get(genre, {})
    mood_kw = d.get("moods", {}).get(mood, {})
    sit_kw = d.get("situations", {}).get(situation, {})

    if language == "ko":
        g = (genre_kw.get("ko_keywords") or [genre])[0]
        m = (mood_kw.get("ko_keywords") or [mood])[0]
        s = (sit_kw.get("ko_keywords") or [situation])[0]
        templates = _OVERLAY_KO_TEMPLATES
    else:
        g = (genre_kw.get("en_keywords") or [genre])[0]
        m = (mood_kw.get("en_keywords") or [mood])[0]
        s = (sit_kw.get("en_keywords") or [situation])[0]
        templates = _OVERLAY_EN_TEMPLATES

    tmpl = random.choice(templates)
    return tmpl.format(genre=g, mood=m, situation=s)


# ── 공개 API ──


def generate_thumbnail(
    analysis: AnalysisResult,
    language: str = "ko",
) -> ThumbnailSuggestion:
    genre = analysis.primary_genre
    mood = analysis.primary_mood
    situation = analysis.primary_situation

    main_kw = _build_main_keywords(genre, mood, situation, language)
    sub_kw = _build_sub_keywords(genre, mood, situation)
    tones = _get_color_tones(genre, mood)
    concepts = _get_visual_concepts(mood, situation)
    has_person = _should_have_person(mood, situation)
    layout = _get_layout(mood)
    overlay = _generate_overlay_text(genre, mood, situation, language)

    return ThumbnailSuggestion(
        main_keywords=main_kw,
        sub_keywords=sub_kw,
        color_tone=tones,
        background_concept=concepts[0] if concepts else "abstract gradient",
        has_person=has_person,
        layout=layout,
        text_overlay=overlay,
    )


def generate_thumbnail_variants(
    analysis: AnalysisResult,
    language: str = "ko",
    count: int = 3,
) -> list[ThumbnailSuggestion]:
    """여러 썸네일 시안 변형을 생성한다."""
    variants: list[ThumbnailSuggestion] = []
    for _ in range(count):
        variants.append(generate_thumbnail(analysis, language))
    return variants
